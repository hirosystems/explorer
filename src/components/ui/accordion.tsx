import { Accordion, HStack } from '@chakra-ui/react';
import { CaretDown } from '@phosphor-icons/react';
import * as React from 'react';

interface AccordionItemTriggerProps extends Accordion.ItemTriggerProps {
  indicatorPlacement?: 'start' | 'end';
}

export const AccordionItemTrigger = React.forwardRef<HTMLButtonElement, AccordionItemTriggerProps>(
  function AccordionItemTrigger(props, ref) {
    const { children, indicatorPlacement = 'end', ...rest } = props;
    return (
      <Accordion.ItemTrigger {...rest} ref={ref}>
        {indicatorPlacement === 'start' && (
          <Accordion.ItemIndicator rotate={{ base: '-90deg', _open: '0deg' }}>
            <CaretDown />
          </Accordion.ItemIndicator>
        )}
        <HStack gap="4" flex="1" textAlign="start" width="full">
          {children}
        </HStack>
        {indicatorPlacement === 'end' && (
          <Accordion.ItemIndicator>
            <CaretDown />
          </Accordion.ItemIndicator>
        )}
      </Accordion.ItemTrigger>
    );
  }
);

interface AccordionItemContentProps extends Accordion.ItemContentProps {}

export const AccordionItemContent = React.forwardRef<HTMLDivElement, AccordionItemContentProps>(
  function AccordionItemContent({ ...rest }, ref) {
    return (
      <Accordion.ItemContent {...rest} ref={ref}>
        <Accordion.ItemBody {...rest} ref={ref} />
      </Accordion.ItemContent>
    );
  }
);

export const AccordionRoot = Accordion.Root;
export const AccordionItem = Accordion.Item;
