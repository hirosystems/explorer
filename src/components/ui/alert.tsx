import { Alert as ChakraAlert } from '@chakra-ui/react';
import * as React from 'react';

import { CloseButton } from './close-button';

export interface AlertProps extends Omit<ChakraAlert.RootProps, 'title'> {
  startElement?: React.ReactNode;
  endElement?: React.ReactNode;
  title?: React.ReactNode;
  description?: React.ReactNode;
  icon?: React.ReactElement;
  closable?: boolean;
  onClose?: () => void;
}

export const Alert = React.forwardRef<HTMLDivElement, AlertProps>(function Alert(props, ref) {
  const { title, description, icon, closable, onClose, startElement, endElement, ...rest } = props;
  return (
    <ChakraAlert.Root ref={ref} {...rest}>
      {startElement || <ChakraAlert.Indicator>{icon}</ChakraAlert.Indicator>}
      <ChakraAlert.Content>
        {title && <ChakraAlert.Title>{title}</ChakraAlert.Title>}
        {description && <ChakraAlert.Description>{description}</ChakraAlert.Description>}
      </ChakraAlert.Content>
      {endElement}
      {closable && (
        <CloseButton
          size="sm"
          pos="relative"
          top="-2"
          insetEnd="-2"
          alignSelf="flex-start"
          onClick={onClose}
        />
      )}
    </ChakraAlert.Root>
  );
});
