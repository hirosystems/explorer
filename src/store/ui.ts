import { atomFamily } from 'jotai/utils';
import { atom } from 'jotai';

export const tabIndexState = atomFamily<string, number, number>(_param => atom(0));
