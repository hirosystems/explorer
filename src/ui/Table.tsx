'use client';

import {
  Table as CUITable,
  TableProps as CUITableProps,
  forwardRef,
  useColorMode,
} from '@chakra-ui/react';

import { UIComponent } from './types';

export type TableProps = CUITableProps & UIComponent;
export const Table = forwardRef<TableProps, 'table'>(({ children, size, ...rest }, ref) => (
  <CUITable
    ref={ref}
    width={size || rest.width}
    height={size || rest.height}
    minWidth={size || rest.minWidth}
    minHeight={size || rest.minHeight}
    {...rest}
  >
    {children}
  </CUITable>
));
