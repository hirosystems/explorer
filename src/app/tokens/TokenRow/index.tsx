import { useColorMode } from '@chakra-ui/react';
import { FtBasicMetadataResponse } from '@hirosystems/token-metadata-api-client';
import React, { FC } from 'react';

import { TokenLink, TxLink } from '../../../common/components/ExplorerLinks';
import { numberToString } from '../../../common/utils/utils';
import { Flex } from '../../../ui/Flex';
import { Td } from '../../../ui/Td';
import { Text } from '../../../ui/Text';
import { Tr } from '../../../ui/Tr';
import { TokenAvatar } from '../../address/[principal]/TokenBalanceCard/TokenAvatar';

export const TokenRow: FC<{
  ftToken: FtBasicMetadataResponse;
}> = ({ ftToken }) => {
  const colorMode = useColorMode().colorMode;
  const name = ftToken.name || 'FT Token';
  // `${ftToken.contract_principal}::${name}`
  return (
    <Tr>
      <Td padding={'10px 20px 10px 16px'} width={['auto', 'auto', '30%']}>
        <Flex alignItems={'center'} gap={'8px'}>
          <TokenAvatar metadataImageUrl={ftToken.image_uri} asset={ftToken.symbol || 'FT'} />
          <TokenLink
            tokenId={ftToken.contract_principal}
            fontSize={'sm'}
            whiteSpace={'nowrap'}
            textOverflow={'ellipsis'}
            overflow={'hidden'}
          >
            {name} {ftToken.symbol ? `(${ftToken.symbol})` : null}
          </TokenLink>
        </Flex>
      </Td>
      <Td padding={'10px'} display={['none', 'none', 'table-cell']}>
        <Text fontSize={'15px'} whiteSpace={'nowrap'} textOverflow={'ellipsis'} overflow={'hidden'}>
          <TxLink txId={ftToken.tx_id}>{ftToken.tx_id}</TxLink>
        </Text>
      </Td>
      <Td isNumeric width={'130px'} padding={'10px 16px 10px 20px'}>
        <Text whiteSpace={'nowrap'} textOverflow={'ellipsis'} overflow={'hidden'} fontSize={'15px'}>
          {numberToString(ftToken.total_supply ? Number(ftToken.total_supply) : 0)}
        </Text>
      </Td>
    </Tr>
  );
};
