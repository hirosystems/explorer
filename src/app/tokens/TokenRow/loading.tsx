import React, { FC } from 'react';

import { Box } from '../../../ui/Box';
import { Flex } from '../../../ui/Flex';
import { SkeletonItem } from '../../../ui/SkeletonItem';
import { Td } from '../../../ui/Td';
import { Tr } from '../../../ui/Tr';

export const Loading: FC = () => {
  return (
    <Tr>
      <Td padding={'10px 20px 10px 16px'} width={['auto', 'auto', '30%']}>
        <Flex alignItems={'center'} gap={'8px'}>
          <SkeletonItem width={'36px'} height={'36px'} />
          <Box width={'100%'} maxWidth={'120px'}>
            <SkeletonItem height={'15px'} />
          </Box>
        </Flex>
      </Td>
      <Td padding={'10px'} display={['none', 'none', 'table-cell']}>
        <Box maxWidth={'500px'}>
          <SkeletonItem height={'15px'} />
        </Box>
      </Td>
      <Td isNumeric width={'130px'} padding={'10px 16px 10px 20px'}>
        <Box maxWidth={'70px'} marginLeft={'auto'}>
          <SkeletonItem height={'15px'} />
        </Box>
      </Td>
    </Tr>
  );
};
