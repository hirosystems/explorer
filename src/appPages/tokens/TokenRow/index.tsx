import { useColorMode } from '@chakra-ui/react';
import { Tr } from '@/ui/Tr';
import { Td } from '@/ui/Td';
import { numberToString } from '@/common/utils';
import { Circle } from '@/ui/Circle';
import { Image } from '@/ui/Image';
import { Flex } from '@/ui/Flex';
import { Text } from '@/ui/typography';
import { TokenLink, TxLink } from '@/components/links';

export function TokenRow({
  symbol,
  imgUrl,
  name,
  totalSupply,
  txId,
  tokenId,
}: {
  name: string;
  txId: string;
  tokenId: string;
  imgUrl?: string;
  symbol?: string;
  totalSupply?: string;
}) {
  const { colorMode } = useColorMode();
  return (
    <Tr>
      <Td padding="10px 20px 10px 16px" width={['auto', 'auto', '30%']}>
        <Flex alignItems="center" gap="8px">
          <Image
            width="36px"
            height="36px"
            src={imgUrl}
            fallback={
              <Circle size="36px" flexShrink="0">
                {name[0].toUpperCase()}
              </Circle>
            }
          />
          <TokenLink tokenId={tokenId}>
            <Text
              as="a"
              fontSize="15px"
              whiteSpace="nowrap"
              textOverflow="ellipsis"
              overflow="hidden"
              color={`links.${colorMode}`}
              cursor="pointer"
            >
              {name} ({symbol})
            </Text>
          </TokenLink>
        </Flex>
      </Td>
      <Td padding="10px" display={['none', 'none', 'table-cell']}>
        <Text
          fontSize="15px"
          whiteSpace="nowrap"
          textOverflow="ellipsis"
          overflow="hidden"
          color={`links.${colorMode}`}
        >
          <TxLink txId={txId}>{txId}</TxLink>
        </Text>
      </Td>
      <Td isNumeric width="130px" padding="10px 16px 10px 20px">
        <Text whiteSpace="nowrap" textOverflow="ellipsis" overflow="hidden" fontSize="15px">
          {numberToString(totalSupply ? Number(totalSupply) : 0)}
        </Text>
      </Td>
    </Tr>
  );
}
