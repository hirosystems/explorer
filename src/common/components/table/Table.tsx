import { ScrollableBox } from '@/app/_components/BlockList/ScrollableDiv';
import { Text } from '@/ui/Text';
import { Tooltip } from '@/ui/Tooltip';
import { Table as ChakraTable, Flex, Icon } from '@chakra-ui/react';
import styled from '@emotion/styled';
import { ArrowDown, ArrowUp, ArrowsDownUp, Info } from '@phosphor-icons/react';
import React, { Suspense, useCallback, useState } from 'react';

import { ExplorerErrorBoundary } from '../../../app/_components/ErrorBoundary';
import { mobileBorderCss } from '../../constants/constants';
import { TableContainer } from './TableContainer';
import { TableSkeleton } from './TableSkeleton';

const StyledTable = styled(ChakraTable.Root)`
  tr td {
    border-bottom: none;
  }
`;

type SortOrder = 'asc' | 'desc';

interface SortIconProps {
  sortable: boolean | undefined;
  sortColumn: string | undefined | null;
  columnId: string;
  sortOrder: SortOrder | undefined;
  onSort: ((columnId: string, direction: SortOrder) => void) | undefined;
}

function SortIcon({ sortable, sortColumn, columnId, sortOrder, onSort }: SortIconProps) {
  if (!sortable) return null;
  return (
    <Icon
      onClick={() => {
        onSort?.(columnId, sortOrder === 'asc' ? 'desc' : 'asc');
      }}
      h={4}
      w={4}
    >
      {sortColumn !== columnId ? (
        <ArrowsDownUp />
      ) : sortOrder === 'asc' ? (
        <ArrowUp />
      ) : (
        <ArrowDown />
      )}
    </Icon>
  );
}

export const TableHeader = <T extends unknown[]>({
  columnDefinition,
  sortColumn,
  sortOrder,
  headerTitle,
  isFirst,
  onSort,
}: {
  sortColumn?: string | null;
  sortOrder?: SortOrder;
  columnDefinition: ColumnDefinition<T>;
  headerTitle: string | React.ReactNode;
  isFirst: boolean;
  onSort?: (columnId: string, direction: SortOrder) => void;
}) => {
  return (
    <ChakraTable.Header
      py={3}
      px={6}
      border="none"
      css={isFirst ? mobileBorderCss : {}}
      width="fit-content"
      position={isFirst ? 'sticky' : 'unset'}
      left={isFirst ? 0 : undefined}
      zIndex={isFirst ? 'docked' : undefined}
      bg="surface"
      borderBottom="1px solid var(--stacks-colors-borderSecondary)"
      role="columnheader"
      aria-sort={sortOrder ? (sortOrder === 'asc' ? 'ascending' : 'descending') : 'none'}
    >
      {typeof headerTitle === 'string' ? ( // TODO: why not also use a custom renderer
        <Flex gap={1.5} alignItems="center" py={4}>
          <Text
            fontWeight="normal"
            whiteSpace="nowrap"
            fontSize="sm"
            color={{
              _light: 'slate.700',
              _dark: 'slate.250',
            }}
            textTransform="none"
            letterSpacing="normal"
            fontFamily="instrument"
          >
            {headerTitle}
          </Text>
          {columnDefinition.tooltip && (
            <Tooltip content={columnDefinition.tooltip}>
              <Icon h={4} w={4}>
                <Info />
              </Icon>
            </Tooltip>
          )}
          <SortIcon
            sortable={columnDefinition.sortable}
            sortColumn={sortColumn}
            columnId={columnDefinition.id}
            sortOrder={sortOrder}
            onSort={onSort}
          />
        </Flex>
      ) : (
        headerTitle
      )}
    </ChakraTable.Header>
  );
};

export function TableRow<T extends unknown[]>({
  rowData,
  columns,
  rowIndex,
  isFirst,
  isLast,
}: {
  rowData: T;
  columns: ColumnDefinition<T>[];
  rowIndex: number;
  isFirst: boolean;
  isLast: boolean;
}) {
  const [isHovered, setIsHovered] = useState(false);
  return (
    <ChakraTable.Row
      _hover={{
        bg: isHovered ? 'sand.150' : 'inherit',
      }}
      onMouseEnter={() => setIsHovered(true)}
      onMouseLeave={() => setIsHovered(false)}
      css={{
        '& > td:first-of-type': {
          borderTopLeftRadius: 'xl',
          borderBottomLeftRadius: 'xl',
        },
        '& > td:last-of-type': {
          borderTopRightRadius: 'xl',
          borderBottomRightRadius: 'xl',
        },
      }}
    >
      {columns.map((col, colIndex) => (
        <ChakraTable.Cell
          key={`table-row-${rowIndex}-col-${colIndex}`}
          py={4}
          px={6}
          css={{
            ...(colIndex === 0
              ? { ...mobileBorderCss, bg: isHovered ? 'inherit' : 'surface' }
              : {}),
          }}
          position={colIndex === 0 ? 'sticky' : 'unset'}
          left={colIndex === 0 ? 0 : undefined}
          zIndex={colIndex === 0 ? 'docked' : undefined}
        >
          {col.cellRenderer ? (
            col.cellRenderer(col.accessor(rowData))
          ) : (
            <Text whiteSpace="nowrap" overflow="hidden" textOverflow="ellipsis" fontSize="sm">
              {col.accessor(rowData)}
            </Text>
          )}
        </ChakraTable.Cell>
      ))}
    </ChakraTable.Row>
  );
}

export interface ColumnDefinition<T extends unknown[], R = string> {
  id: string;
  header: string | React.ReactNode;
  tooltip?: string;
  accessor: (row: T) => R;
  sortable?: boolean;
  onSort?: (a: T, b: T) => number;
  cellRenderer?: (value: R) => React.ReactNode;
}

export interface TableProps<T extends unknown[]> {
  title?: string;
  topRight?: React.ReactNode;
  topLeft?: React.ReactNode;
  rowData: T[];
  columnDefinitions: ColumnDefinition<T>[];
}

export function Table<T extends unknown[]>({
  title,
  topRight,
  topLeft,
  rowData,
  columnDefinitions,
}: TableProps<T>) {
  const [sortColumnId, setSortColumnId] = useState<string | undefined>(undefined);
  const [sortDirection, setSortDirection] = useState<SortOrder | undefined>(undefined);

  const [sortedRowData, setSortedRowData] = useState(rowData);

  const onSort = useCallback(
    (columnId: string, sortOrder: SortOrder) => {
      const columnDefinition = columnDefinitions.find(col => col.id === columnId);
      if (!columnDefinition) {
        throw new Error(`Column definition not found for columnId: ${columnId}`);
      }
      if (!columnDefinition.sortable) {
        throw new Error(`Column ${columnId} is not sortable`);
      }
      if (!columnDefinition.onSort) {
        throw new Error(`Column ${columnId} does not have an onSort function`);
      }
      setSortColumnId(columnId);
      setSortDirection(sortOrder);
      setSortedRowData(rowData.sort(columnDefinition.onSort));
    },
    [columnDefinitions, rowData]
  );

  return (
    <ExplorerErrorBoundary Wrapper={TableContainer} tryAgainButton>
      <Suspense
        fallback={
          <TableSkeleton numColumns={columnDefinitions.length} numRows={rowData.length ?? null} />
        }
      >
        {/* <TableProvider initialData={data}> */}
        <TableContainer topRight={topRight} topLeft={topLeft} title={title}>
          <ScrollableBox>
            <StyledTable width="full">
              <ChakraTable.Header>
                <ChakraTable.Row>
                  {columnDefinitions?.map((col, colIndex) => (
                    <TableHeader
                      key={col.id}
                      columnDefinition={col}
                      headerTitle={col.header}
                      sortColumn={sortColumnId}
                      sortOrder={sortDirection}
                      isFirst={colIndex === 0}
                      onSort={onSort}
                    />
                  ))}
                </ChakraTable.Row>
              </ChakraTable.Header>
              <ChakraTable.Body>
                {sortedRowData?.map((rowData, rowIndex) => (
                  <TableRow
                    key={rowIndex}
                    rowIndex={rowIndex}
                    rowData={rowData}
                    columns={columnDefinitions}
                    isFirst={rowIndex === 0}
                    isLast={rowIndex === sortedRowData.length - 1}
                  />
                ))}
              </ChakraTable.Body>
            </StyledTable>
          </ScrollableBox>
        </TableContainer>
        {/* </TableProvider> */}
      </Suspense>
    </ExplorerErrorBoundary>
  );
}
