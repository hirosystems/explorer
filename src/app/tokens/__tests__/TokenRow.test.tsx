import renderer from 'react-test-renderer';

import { Table } from '../../../ui/Table';
import { TokenRow } from '../TokenRow';

const tokenMetadata = {
  name: 'TestToken',
  tx_id: '0x1234',
  contract_principal: '0x5678',
  symbol: 'TT',
  total_supply: '1000000',
  image_uri: 'https://example.com/token.png',
  sender_address: '0x1234',
};

describe('TokenRow', () => {
  it('renders correctly', () => {
    const tree = renderer
      .create(
        <Table>
          <TokenRow ftToken={tokenMetadata} />
        </Table>
      )
      .toJSON();
    expect(tree).toMatchSnapshot();
  });
});
