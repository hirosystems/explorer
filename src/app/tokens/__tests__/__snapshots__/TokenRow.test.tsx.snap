// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TokenRow renders correctly 1`] = `
<table
  className="chakra-table css-0"
>
  <tr
    className="css-0"
  >
    <td
      className="css-15d3wjr"
    >
      <div
        className="css-axw7ok"
      >
        <img
          alt="TT"
          height={36}
          onError={[Function]}
          src="https://example.com/token.png"
          style={
            Object {
              "marginRight": "16px",
            }
          }
          width={36}
        />
        <a
          className="chakra-link css-hygcjl"
          href="/token/0x5678?chain=mainnet"
        >
          TestToken
           
          (TT)
        </a>
      </div>
    </td>
    <td
      className="css-55yph6"
    >
      <span
        className="chakra-text css-q62vf"
      >
        <a
          className="chakra-link css-0"
          href="/txid/0x1234?chain=mainnet"
        >
          0x1234
        </a>
      </span>
    </td>
    <td
      className="css-1ttiw4m"
      data-is-numeric={true}
    >
      <span
        className="chakra-text css-ae4fie"
      >
        1.00M
      </span>
    </td>
  </tr>
</table>
`;
